### PR TITLE
CumConcatLayer

### PR DESCRIPTION
**Pull request by albertz**
This is for #391.

### CODE REVIEW

File: returnn/tf/util/data.py [hunk 1]
```diff
@@ -2731,15 +2735,25 @@ def get_axes_from_description(self, axes, allow_int=True):
...
     :param bool allow_int:
     :return: axis, counted with batch-dim
     :rtype: int
     """
     axes = self.get_axes_from_description(axis, allow_int=allow_int)
+    assert axes, "%s: %r axis not found" % (self, axis)
     assert len(axes) == 1, "%r: %r is not a unique axis but %r" % (self, axis, axes)
     return axes[0]
 
+  def has_axis(self, axis):
```
[the remark this answers is not in the document]
Comment (Zettelkasten) [reply returnn/tf/util/data.py]: (Sorry for my late answer. I see you merged now, wasn't too important.)
This was about the method name. I thought of this primarily as a way of checking that `get_axis_from_description(self, axis: str|DimensionTag)` would not raise an error.
I understand your reasoning though, and yes, I agree. I dislike a little that now both dimension tags and `int` axes are called `axis` in the code, maybe `has_dim_tag(self, dim_tag: DimensionTag|str)` would make make a little more sense? Dim tags and axes are conceptually very similar, but not entirely the same, e.g. different data can have the same dim tags at different axes.

**Comment by albertz** [comment]
This turns out to touch on a lot of previous somewhat hacky code, which breaks in many different ways now, every time something got changed...

The changes are to better handle the dyn seq len inside a dim tag w.r.t. different batches and beams.

Once this is ready, I will probably move this all out to a separate PR, and also try to clean it up, and maybe split it up. With ready I mean excluding the `CumConcatLayer` and changes in `DotLayer` and the new tests for those.

It's still not ready currently but I think soon...

**Comment by albertz** [comment]
Ok, finally, all tests are passing, except the new ones.
Now cleaning up...

**Comment by albertz** [comment]
I wonder a bit how to best handle such huge PRs (e.g. how to effectively do the cleanup).
I wrote that down [here on Reddit as a question](https://www.reddit.com/r/learnprogramming/comments/pftenx/how_to_cleanup_a_branch_pr_with_huge_number_of/) but not sure if that is the best place to ask such things. But I also don't really know any good place where I could ask such question. Any recommendation where to ask such things?

**Comment by albertz** [comment]
I will probably now first move everything out into a new branch, except the two new test cases and also the `CumConcatLayer` (which is what this PR here is about + further things we need for the test cases to pass). Because all the remaining things should always pass all the tests.

**Comment by albertz** [comment]
Ok, all the other changes are now in this branch: https://github.com/rwth-i6/returnn/tree/albert-generalized-self-att-wip-2021-09-01.

**Comment by albertz** [comment]
Wow, this was quite some effort, to get the basics ready, to continue working on this itself.

Basically:
#592 #599 #600 #601 #602 #603 #604 #605 #606 #607 #608 #609 #610 #611 #612 #613 #614 #615 #616 #617 #618 #619 #620 #621 #622 #623 #624

**Comment by albertz** [comment]
The test case is a full simple test case for auto-regressive self attention, so it covers #391. `CumConcatLayer` is the final missing piece now, so this finally fixes #391 (at least the basic support).

**Pull request merged**